### ISSUE TITLE
Kotlin QuarkusMain native build does not work

### ISSUE THREAD

**Issue by arolfes** [quarkusio/quarkus#29754]
### Describe the bug

Hello,

I have a main application class and want to compile everything in native mode.

### Main Class

```kotlin
package com.github

import io.quarkus.runtime.Quarkus.run
import io.quarkus.runtime.annotations.QuarkusMain

@QuarkusMain
class GreetingApplication

fun main(args: Array<String>) {
    run(*args)
}
```

during initializing phase it breaks with the following exception:

```
./gradlew build -Dquarkus.package.type=native

.
.
.

========================================================================================================================
GraalVM Native Image: Generating 'quarkus-kotlin-native-1.0.0-SNAPSHOT-runner' (executable)...
========================================================================================================================
[1/7] Initializing...                                                                                    (0.0s @ 0.15GB)
Error: Method 'com.github.GreetingApplication.main' is declared as the main entry point but it can not be found. Make sure that class 'com.github.GreetingApplication' is on the classpath and that method 'main(String[])' exists in that class.
com.oracle.svm.core.util.UserError$UserException: Method 'com.github.GreetingApplication.main' is declared as the main entry point but it can not be found. Make sure that class 'com.github.GreetingApplication' is on the classpath and that method 'main(String[])' exists in that class.
 
```

### Expected behavior

native build works also with a kotlin main class

### Actual behavior

Build breaks during Initialization Phase with this error message:

[1/7] Initializing...                                                                                    (0.0s @ 0.15GB)
Error: Method 'com.github.GreetingApplication.main' is declared as the main entry point but it can not be found. Make sure that class 'com.github.GreetingApplication' is on the classpath and that method 'main(String[])' exists in that class.
com.oracle.svm.core.util.UserError$UserException: Method 'com.github.GreetingApplication.main' is declared as the main entry point but it can not be found. Make sure that class 'com.github.GreetingApplication' is on the classpath and that method 'main(String[])' exists in that class.

### How to Reproduce?

1) clone the sample repo: https://github.com/arolfes/quarkus-kotlin-native
2) run nativeBuild `./gradlew build -Dquarkus.package.type=native`

### Output of `uname -a` or `ver`

Linux fedora 6.0.11-300.fc37.x86_64 #1 SMP PREEMPT_DYNAMIC Fri Dec 2 20:47:45 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux

### Output of `java -version`

OpenJDK Runtime Environment GraalVM CE 22.3.0 (build 17.0.5+8-jvmci-22.3-b08)

### Quarkus version or git rev

2.14.3.Final

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Gradle 7.5.1

**Comment by geoand** [comment]
Thanks for reporting.

https://github.com/quarkusio/quarkus/pull/29755 fixes the issue

**Comment by arolfes** [comment]
That was fast.
thank you

**Comment by geoand** [comment]
👍

**Comment by evanchooly** [comment]
Does this work in JVM mode?  because `GreetingApplication` is an empty class and contains no functions.  that `main()` belongs to a class called `<FileName>kt`.  If the file name just happens to be `GreetingApplication.kt` then perhaps you're getting this association for free but I would consider a class set up this way to be a bug in the application itself.  (I haven't seen @geoand's PR yet so maybe the fix is simple.)

**Comment by arolfes** [comment]
Hello @evanchooly ,

the problem exists also in the normal run mode. you can't start.

```
2022-12-08 19:48:41,477 INFO  [io.qua.dep.dev.IsolatedDevModeMain] (main) Attempting to start live reload endpoint to recover from previous Quarkus startup failure
> :qu2022-12-08 19:48:41,725 ERROR [io.qua.dep.dev.IsolatedDevModeMain] (main) Failed to start quarkus: java.lang.NoSuchMethodException: com.github.GreetingApplication.main([Ljava.lang.String;)
        at java.base/java.lang.Class.getMethod(Class.java:2227)
        at io.quarkus.runner.bootstrap.StartupActionImpl.runMainClass(StartupActionImpl.java:98)
        at io.quarkus.deployment.dev.IsolatedDevModeMain.firstStart(IsolatedDevModeMain.java:137)
        at io.quarkus.deployment.dev.IsolatedDevModeMain.accept(IsolatedDevModeMain.java:447)
        at io.quarkus.deployment.dev.IsolatedDevModeMain.accept(IsolatedDevModeMain.java:59)
        at io.quarkus.bootstrap.app.CuratedApplication.runInCl(CuratedApplication.java:149)
        at io.quarkus.bootstrap.app.CuratedApplication.runInAugmentClassLoader(CuratedApplication.java:104)
        at io.quarkus.deployment.dev.DevModeMain.start(DevModeMain.java:131)
        at io.quarkus.deployment.dev.DevModeMain.main(DevModeMain.java:62)
```

@geoand : Does your commit fix this problem also? should I open a new bug report?

**Comment by evanchooly** [comment]
That's precisely my point.  This `main()` doesn't exist on `GreetingApplication`.  At best it exists on `GreetingApplicationKt` becaue the `main()` is a top level function.  If you want this precise source to work try updating your annotation like so:  `@file:QuarkusMain`.  That will apply the annotation to the file and thus, potentially, the `GreetingApplicationKt` class synthesized to hold your top level function.

In short, this is not a quarkus bug and should not be fixed in quarkus.  This is a function of how kotlin works and thus not a bug at all but a misunderstanding how where elements live and how to annotate them properly.

**Comment by geoand** [comment]
My PR should address all cases

**Comment by geoand** [comment]
@evanchooly I don't agree that this should not be fixed.

Users are used to doing something like:

```
@SpringBootApplication
class DemoApplication

fun main(args: Array<String>) {
    runApplication<DemoApplication>(*args)
}
```

because that's how they are told to do it in Spring Boot. We should seamlessly support this same intuitive pattern. My PR does take care of it, while also allowing for the more verbose syntax with the companion object.

**Comment by evanchooly** [comment]
If that's how Spring does it then perhaps being consistent with that has its merits.  Spring made the wrong choice there, IMO, as this changes the semantics of the language.
